OpenZFS 4185 - add new cryptographic checksums to ZFS: SHA-512, Skein, Edon-R

tonyhutter opened the pull request:
--------
OpenZFS 4185 - add new cryptographic checksums to ZFS: SHA-512, Skein, Edon-R

Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: Prakash Surya <prakash.surya@delphix.com>
Reviewed by: Saso Kiselkov <saso.kiselkov@nexenta.com>
Reviewed by: Richard Lowe <richlowe@richlowe.net>
Approved by: Garrett D'Amore <garrett@damore.org>
Ported by: Tony Hutter <hutter2@llnl.gov>

OpenZFS-issue: https://www.illumos.org/issues/4185
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/45818ee

Porting Notes:
This code is ported on top of the Illumos Crypto Framework code:

    https://github.com/zfsonlinux/zfs/pull/4329/commits/b5e030c8dbb9cd393d313571dee4756fbba8c22d

Porting changes include:

- Copied full `module/icp/include/sha2/sha2.h` directly from illumos

- Removed from `module/icp/algs/sha2/sha2.c`:
	`#pragma inline(SHA256Init, SHA384Init, SHA512Init)`

- Added 'ctx' to `lib/libzfs/libzfs_sendrecv.c:zio_checksum_SHA256()` since
  it now takes in an extra parameter.

- Added CTASSERT() to assert.h from for `module/zfs/edonr_zfs.c`

- Added skein & edonr to `libicp/Makefile.am`

- Added `sha512.S`.  It was generated from `sha512-x86_64.pl` in Illumos.

- Updated `ztest.c` with new `fletcher_4_*()` args; used NULL for new ctx argument.

- In `icp/algs/edonr/edonr_byteorder.h`, Removed the` #if defined(__linux)` section
  to not #include the non-existent `endian.h`.

- Fixup test files:
	- Rename `<sys/varargs.h>` ->` <varargs.h>`, `<strings.h>` -> `<string.h>`,
	- Remove <note.h> and define `NOTE(`) as NOP.
	- Define `u_longlong_t`
	- Rename "`#!/usr/bin/ksh`" -> "`#!/bin/ksh -p`"
	- Rename NULL to 0 in "no test vector" array entries to get around a
	  compiler warning.
	- Remove "for isa in $($ISAINFO); do" stuff
	- Add/update Makefiles
	- Add some userspace headers like stdio.h/stdlib.h in places of
	  sys/types.h.

- EXPORT_SYMBOL `*_Init()/*_Update()/*_Final()...` routines in ICP modules.

- Update `scripts/zfs2zol-patch.sed`

- include <sys/sha2.h> in sha2_impl.h
    
- Add sha2.h to include/sys/Makefile.am
    
- Add skein and edonr dirs to icp Makefile
    
- Add new checksums to zpool_get.cfg
    
- Move checksum switch block from zfs_secpolicy_setprop() to zfs_check_settable()
    
- Fix -Wuninitialized error in edonr_byteorder.h on PPC
    
- Fix stack frame size errors on ARM32
        - Don't unroll loops in Skein on 32-bit to save stack space
        - Add memory barriers in sha2.c on 32-bit to save stack space
    
- Add filetest_001_pos.ksh checksum sanity test
  
- Add option to write psudorandom data in file_write utility

behlendorf commented:
Add the follow line to the commit message for the top most patch in this patch stack.  The buildbot will then pull that SPL commit for testing.

```
Requires-spl: refs/pull/561/head
```

behlendorf commented:
@tonyhutter looking good.  Can you squash your various fixes and rebase this against the latest master.

tonyhutter commented:
@behlendorf will do.  I also need to sanity test it on my PPC and ARM VMs as well.

behlendorf commented:
@tonyhutter can you rebase and squash this in to a stack of patches which can be merged.

tonyhutter commented:
I'm waiting for the first patch of https://github.com/zfsonlinux/zfs/pull/4329 to get pulled first (I just marked it LGTM).   It will be easier to rebase once that patch is in since my code depends on it.

behlendorf commented:
@tonyhutter ahh OK.  Then let me get that merged right away and I'll add your signed off.  It LGTM as well.

behlendorf commented:
OK, the icp module has been merged to master.

tonyhutter commented:
I've re-based the patch on top of ICP, and have begun work on an additional, real-world test to satisfy my paranoia about the new checksum code.

lundman commented:
Oh yes, there were some minor things that we pushed  upstream, as well as, coming from upstream. 
Check you also get;

https://github.com/openzfsonosx/zfs/commit/bef06e165c5d11556451d9c9da69802cb78e005d
https://github.com/openzfsonosx/zfs/commit/b62a65274ab591fda019189059f41bf4304e893e

hope it helps.

behlendorf commented:
@lundman definitely should help, thanks for the heads up!

tonyhutter commented:
I just included the sanity test.  Still need to check if I got the two changes that @lundman pointed out.

behlendorf commented:
Resubmitted the test runs which failed.  Only the  CentOS 7.1 failure might have been legitimately related to this change but there wasn't a console back trace to say for certain.

@tonyhutter assuming the test runs pass this time are you OK with this getting merged?

behlendorf commented:
@lundman @tcaputi and additional reviews you can offer on this change before it's merged would be welcome.

tonyhutter commented:
@behlendorf  I'm ok with it going in.  As you mentioned, getting some reviewer's eyes on it would be good.

tcaputi commented:
@behlendorf ill try to look at it tonight

tonyhutter commented:
Thanks all.  I also just pushed a small commit (https://github.com/zfsonlinux/zfs/pull/4760/commits/ef6f3641f90652baa0aeeffbc1094b8a26583fa1) that changes my checksum tests to use a 100MB pool rather than the default 2GB pool.  Speeds up the test quite a bit!

lundman commented:
LGTM.

tcaputi commented on module/icp/algs/edonr/edonr.c:
```diff
@@ -0,0 +1,751 @@
+ * the function will be called from a worker thread that won't be using
+ * much stack.  The only function that goes over the 1k limit is Q512(),
+ * which only goes over it by a hair (1248 bytes on ARM32).
+ */
```
@behlendorf Can we get a fact check here just to be safe? I'd hate to change any of the crypto code for something like this, but I don't want to break anything on a specific architecture either.

tcaputi replied:
After looking at [this ztest failure](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20%28TEST%29/builds/2075/steps/shell_5/logs/stdio) it looks like this might actually be a problem. Ztest has been known to fail randomly for other reasons, though so this might not actually be the problem.

tonyhutter replied:
Note that the small stack frame size is only a requirement for kernel code, not userspace. 32-bit and some other archs add a warning if a kernel stack frame is over 1k; x86-64 is 2k.  Userspace really has no restrictions on stack size, other than what ulimit sets for it.  I don't know enough about the ztest failure to know if it's a stack frame overflow or not.  

I did talk with @behlendorf about this offline, and we decided it was better to disable the warning in this case since the stack frame was close to 1k, and because it would have been riskier to re-write the function to use less stack.  The low-level checksum code is really hairy stuff.  It uses a ton of local variables for performance, and can't be broken up easily.

behlendorf replied:
Right, this felt like the safest approach to resolving this issue for 32-bit systems.  The ztest failure you linked to I believe is a long standing issue we see occasionally with pthreads which hasn't been run down yet.

tcaputi replied:
OK. Fair enough.

tcaputi commented on module/icp/algs/sha2/sha2.c:
```diff
@@ -336,22 +686,86 @@ SHA2Init(uint64_t mech, SHA2_CTX *ctx)
+		ctx->state.s64[6] = 0x2B0199FC2C85B8AAULL;
+		ctx->state.s64[7] = 0x0EB72DDC81C52CA2ULL;
+		break;
+#ifdef _KERNEL
```
`cmn_err()` is defined in libzpool's kernel.c We should probably leave this in for debugging checksums in userspace

behlendorf replied:
Good catch, this looks like it is unmodified from the illumos source.  But since we can build it in user and kernel space you're right we should.  We'll need to included the header above as well.

tcaputi commented on module/icp/algs/skein/THIRDPARTYLICENSE:
```diff
@@ -0,0 +1,3 @@
+Implementation of the Skein hash function.
+Source code author: Doug Whiting, 2008.
+This algorithm and source code is released to the public domain.
```
When I ported the ICP originally, there was some talk about not being able to use public domain code because it isn't recognized in some places (like France?). If it's in Illumos though its probably fine.

rlaager replied:
Technically, the author should use a real license grant. It's (probably; there is a minority of experts who disagree) not possible to voluntarily place code in the public domain in the US. In practice, it's probably fine.

behlendorf replied:
I agree this is fine, we're using it under the same terms as illumos.

tcaputi replied:
Works for me.

tcaputi commented on module/icp/io/edonr_mod.c:
```diff
@@ -0,0 +1,63 @@
+};
+
+int
+_init(void)
```
This needs a `_fini()` function. The Illumos kernel sometimes omits these for modules it doesn't expect to be unloaded. The ICP and ZFS however are unloadable and so they need to be destroyable. Luckily this is fairly easy. Lok at sha2_mod.c for an example (you will note it doesn't have a `_fini()` in Illumos either).

tonyhutter replied:
Ok, will add

tcaputi commented on module/icp/io/skein_mod.c:
```diff
@@ -0,0 +1,830 @@
+ * Performs an Update on a context with mblk input data.
+ */
+static int
+skein_digest_update_mblk(skein_ctx_t *ctx, crypto_data_t *data)
```
I did not port mblk functionality for the ICP and ZFS doesn't use it. I delete that code wherever I found it. Again, I think this is something we should either support completely or not at all.

behlendorf replied:
That sounds reasonable.  It makes make think we should in a follow up patch remove all this needless `mod_install/mod_remove` complexity.  We're building all of these algorithms in to the icp module already.  The rest of the ZoL code has already simply replaces the Solaris module infrastructure with what's used under Linux and it hasn't posed any real maintenance burden.  This is the kind of thing which doesn't change much.

tcaputi replied:
@behlendorf I think you meant to put this on the comment above? The only issue is that the `mod_install()` / `mod_remove()` code is actually what tells the ICP to make the algorithms available through the generic interface. Maybe we want to just remove the generic interface and use the underlying algorithms directly, but as you said that work could be done in a later patch.

tonyhutter replied:
> CRYPTO_DATA_MBLK format is useful in networking code where mblk_t structures are common (https://blogs.oracle.com/yenduri/entry/how_to_use_crypto_api)

Yea, probably don't need it :-)  Will remove

tcaputi commented on module/icp/io/edonr_mod.c:
```diff
@@ -0,0 +1,63 @@
+}
+
+int
+_info(struct modinfo *modinfop)
```
This function isn't needed. Linux doesn't have an equivalent for its kernel modules (and so neither does the ICP).

behlendorf replied:
Just FYI, there is the MODULE_INFO() macro which is similar in spirit.

tcaputi replied:
Oh really. My fault. I've never seen that before. Either way, the ICP modules aren't Linux kernel modules so this won't work here. We could add some info code for the ICP as a whole though... but again probably best left for later.

tonyhutter replied:
I'll remove _init/_info.

tcaputi replied:
You should probably keep _init() and add a _fini(). See my comment above.

tcaputi commented on module/icp/io/skein_mod.c:
```diff
@@ -0,0 +1,830 @@
+}
+
+int
+_init(void)
```
See edonr_mod.c comments about `_fini()` and `info()`

tcaputi commented on module/zfs/dmu.c:
```diff
@@ -1767,7 +1770,7 @@ dmu_write_policy(objset_t *os, dnode_t *dn, int level, int wp, zio_prop_t *zp)
 		 */
 		compress = ZIO_COMPRESS_OFF;
-		checksum = ZIO_CHECKSUM_OFF;
+		checksum = ZIO_CHECKSUM_NOPARITY;
```
Will this break backwards compatibility with anything if the ZIO_CHECKSUM_OFF enum is recorded anywhere?

behlendorf replied:
It will, we can't do this until we port the multi vdev crash feature.  We haven't bothered to port that feature to ZoL because it doesn't benefit the Linux port in any way.  The only reason to do it is to avoid issues like this.

tonyhutter replied:
Good catch, I'll switch it back to ZIO_CHECKSUM_OFF.

tcaputi commented on module/zfs/zio_checksum.c:
```diff
@@ -146,6 +223,30 @@ zio_checksum_label_verifier(zio_cksum_t *zcp, uint64_t offset)
+	if (spa->spa_cksum_tmpls[checksum] != NULL)
+		return;
+
+	VERIFY(ci->ci_tmpl_free != NULL);
```
the `VERIFY()`s should probably be `ASSERT()`s. ASSERTS only get compiled in with --debug-enabled

behlendorf replied:
All true, but this isn't performance critical code and upstream OpenZFS used a VERIFY here so let's leave it a VERIFY to avoid introducing additional discrepancies between the code bases.

tcaputi commented on module/zfs/zio_checksum.c:
```diff
@@ -179,10 +283,12 @@ zio_checksum_compute(zio_t *zio, enum zio_checksum checksum,
 			bp->blk_cksum = eck->zec_cksum;
 		eck->zec_magic = ZEC_MAGIC;
-		ci->ci_func[0](data, size, &cksum);
+		ci->ci_func[0](data, size, spa->spa_cksum_tmpls[checksum],
```
I'm not sure this is correct. Your template initialization functions seem to initialize a context, not a context template. If my understanding is correct, the context should be modified by the checksum functions, leaving it invalid for future uses. I'm not seeing where the template gets used to make a new context anywhere....

tonyhutter replied:
This is straight from https://github.com/openzfs/openzfs/commit/45818ee so I don't know if it's correct or not.  The code is still there in top-of-the-tree Illumos.

tonyhutter replied:
Looks like the `ci->ci_func`  got expanded to include the template along with the checksum:
```
typedef void zio_checksum_t(const void *data, uint64_t size,
    const void *ctx_template, zio_cksum_t *zcp);
```

tcaputi commented on module/zfs/zio_checksum.c:
```diff
@@ -146,6 +223,30 @@ zio_checksum_label_verifier(zio_cksum_t *zcp, uint64_t offset)
+
+	if (ci->ci_tmpl_init == NULL)
+		return;
+	if (spa->spa_cksum_tmpls[checksum] != NULL)
```
is there a reason we can't initialize these at spa import / creation time? It would save the zio layer from having to call this function repeatedly just to return instantly.

behlendorf replied:
Good question.  I don't see why this couldn't/shouldn't be done at spa_init() time.  However, since this a port of an existing illumos patch I think we should bring over the code as is, unless it's clearly wrong, and then followup with optimizations.  That will make it easier to submit this stuff back upstream.

tcaputi commented on module/zfs/edonr_zfs.c:
```diff
@@ -0,0 +1,103 @@
+
+	ASSERT(ctx_template != NULL);
+	bcopy(ctx_template, &ctx, sizeof (ctx));
+	EdonRUpdate(&ctx, buf, size * 8);
```
This is fine for now. In the future we might want to try to get everything to use the ICP's generic layer, but seeing as we don't have all of the checksum algorithms in there now this should be fine for now.

tcaputi commented on module/zfs/spa.c:
```diff
@@ -2679,6 +2680,19 @@ spa_load_impl(spa_t *spa, uint64_t pool_guid, nvlist_t *config,
 	}
 
+	/* Grab the secret checksum salt from the MOS. */
+	error = zap_lookup(spa->spa_meta_objset, DMU_POOL_DIRECTORY_OBJECT,
```
This isn't really secret. Any user with sudo can just look this up with zdb, in fact. I would not be comfortable using this as a MAC while the 'key' is so easy to figure out.

behlendorf replied:
Agreed.  Secret is a bad choice of words.  There's nothing secret about it.

tonyhutter replied:
I'll remove "secret" from the comment

tcaputi commented on module/zfs/zfs_ioctl.c:
```diff
@@ -3893,6 +3889,47 @@ zfs_check_settable(const char *dsname, nvpair_t *pair, cred_t *cr)
+		if ((err = spa_open(dsname, &spa, FTAG)) != 0)
+			return (err);
+		/*
+		 * Salted checksums are not supported on root pools.
```
Why not, out of curiosity?

rlaager replied:
I'm not at all sure, but possibly lack of GRUB support?

behlendorf replied:
That's my guess as well.  They certainly could be supported but it'll require GRUB be updated.

tcaputi commented on module/zfs/zio_checksum.c:
```diff
@@ -58,28 +60,102 @@
-	{{fletcher_4_native,	fletcher_4_byteswap},	1, 0, 0, "fletcher4"},
-	{{zio_checksum_SHA256,	zio_checksum_SHA256},	1, 0, 1, "sha256"},
-	{{fletcher_4_native,	fletcher_4_byteswap},	0, 1, 0, "zilog2"},
+	{{NULL, NULL}, NULL, NULL, 0, "inherit"},
```
@behlendorf This table is really hard to read. Can we use a /* BEGIN CSTYLED */ block here? Or maybe we can initialize everything with the `.member_name` syntax. Maybe its just me and everyone else can read it fine.

behlendorf replied:
It's not just you.  But I think we can live with it.

tcaputi commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -147,12 +147,14 @@ function default_setup_noexit
 			destroy_pool $TESTPOOL
 		fi
 		[[ -d /$TESTPOOL ]] && $RM -rf /$TESTPOOL
+		log_note creating poool $TESTPOOL $disklist
```
poool

tonyhutter replied:
Will fix

tcaputi commented on tests/zfs-tests/tests/functional/checksum/edonr_test.c:
```diff
@@ -0,0 +1,219 @@
+		uint8_t		digest[mode / 8];			\
+		EdonRInit(&ctx, mode);					\
+		EdonRUpdate(&ctx, (const uint8_t *) _m, strlen(_m) * 8);\
+		EdonRFinal(&ctx, digest);				\
```
This doesn't test the code the way we are using it in the kernel module. I bet that if you were to use the logic with the reused 'context template' it would fail to work correctly.

tonyhutter replied:
That test is taken from Illumos.  I was a little leary of it too, so I also included a real-world test of the checksums in:
`tests/zfs-tests/tests/functional/checksum/filetest_001_pos.ksh`
Does it look like it will test the context template issue you pointed out?

tcaputi replied:
If i understand that test correctly, I don't think it will. I think it will end up starting with the same state and it will re-evaluate the checksums in the same order it wrote them, so even if it isn't doing the right computation, it is doing the same computation and will end up with the same checksum. I bet if we were to write 2 files and read them in reverse order we wrote them the test would fail. Maybe I am wrong about this, but I am 95% sure there is a problem here.

tonyhutter replied:
@tcaputi  just so I'm clear, are you referring to `edonr_test.c` or `filetest_001_pos.ksh` (which tests the kernel checksum code) as the test that's not going to work correctly?

tcaputi replied:
`filetest_001_pos.ksh`. `edonr_test.c` uses the init() ->update() -> final() pattern correctly where as the kernel code tries to reuse the result of init() after other users have called final() on it. Like I said, I could be wrong but you should have to copy the context for each new buffer you are looking to checksum.

tcaputi commented on tests/zfs-tests/tests/functional/checksum/sha2_test.c:
```diff
@@ -0,0 +1,265 @@
+
+/*
+ * Test messages from:
+ * http://csrc.nist.gov/groups/ST/toolkit/documents/Examples/SHA_All.pdf
```
This test is really good to have in here. Thanks for including it.

tcaputi commented:
@tonyhutter @behlendorf 

First off thank you for implementing these changes. Specifically thank you for re-adding a lot of the sha2 code I took out when implementing sha256 and for adding NIST checks for anyone looking to make these algorithms faster in the future.

I have added some comments to the code above. I also have some general comments below. Some of them are a bit nit-picky and for some of them, I deffer to @behlendorf :

- I see that you have moved a lot of the headers from the ICP's include/ to the global projects. I did this originally to add some separations of concerns, like how the spl has its own include/ directory. Perhaps @behlendorf has some wisdom here from the eventual goals of moving the spl into the zfs project.
- 	I also see some `#ifdef`s for sparc code. I was taking those out while porting the ICP originally because we do not have any architecture specific support for it anywhere else. Do we want to support it for real or leave it to use the generic code.
- 	Do we want to have support for noparity? I feel like this could be misleading for users. They might look at this and say 'raid already gives me bitrot protection so I don't need any checksumming' and turn this on without understanding the implications.
-  We should be adding calls to the `_init()` and `_fini()` functions for each of the new modules to `module/icp/illumos_crypto.c` This will allow the code to be used from the generic ICP layer. This will require renaming these functions to avoid conflicts.

behlendorf commented:
> I see that you have moved a lot of the headers from the ICP's include/ to the global projects. I did this originally to add some separations of concerns, like how the spl has its own include/ directory. Perhaps @behlendorf has some wisdom here from the eventual goals of moving the spl into the zfs project.

They headers could definitely stand to be reorganized so make things simpler.  This probably means moving all the `_impl.h` headers under modules/ and the rest under the global includes.  This simplifies providing `-devel` packages by keeping all the public headers in one spot.  The libspl case is a little bit more complicated because many of those headers conflict with existing libc headers.

Bottom line I think it's OK to add them were you did and someone should take a step back and look at the big picture regarding the layout.  The best time to do this would be when we pull the spl in to the zfs tree.

> I also see some #ifdefs for sparc code. I was taking those out while porting the ICP originally because we do not have any architecture specific support for it anywhere else. Do we want to support it for real or leave it to use the generic code.

So believe it or not we do have some sparc64 users.  However, I think it's probably best to drop the sparc64 specific code because we simply aren't set up to test it.  Forcing them to use the generic implementation at least means they're running code which has been tested somewhere.

> Do we want to have support for noparity?

No we don't so we should really drop this functionality until we do.  This was part of the raidz crash dump patches which were never ported to ZoL because they don't really offer us anything on Linux.  That said, we may want to do this for the sake of portability alone.  But at the moment it's not supported.

We should at a minimum drop the `dmu_write_policy()` change until this is supported.  Even though this should technically be dead code at the moment since Linux handles crash dumps entirely differently than illumos.

> We should be adding calls to the _init() and _fini() functions for each of the new modules to module/icp/illumos_crypto.c This will allow the code to be used from the generic ICP layer. This will require renaming these functions to avoid conflicts.

Absolutely.  @tcaputi thanks for for speedy review on this!

tonyhutter commented:
@tcaputi thanks for the comments.  Let me see if I can answer some of your questions:
- I can't remember 100% why I moved some of the headers (it was months ago when I started on this...) but I think part of the reason was so they could be easily #included in the automated tests (like sha2_test.c).
- The `#if defined(__sparc)` stuff is from Illumos.  I'm hesitant to remove it because we also don't know if the generic implementation works under sparc.  At least the sparc-specific code has been tested on somebody's sparc box somewhere at some point in time.  Very reassuring, I know...
- I'll keep `noparity` in for completeness, with the exception of the `dmu_write_policy()` issue you pointed out.  In regard to users using `noparity` on their pools, the man page does warn:

> The value noparity  not only  disables integrity but also disables maintaining parity for user data.  This setting is used internally by a dump device residing on a RAID-Z pool and should not be used by any other dataset.  Disabling checksums is NOT a recommended practice.

- I'll add the `*_init()/*_fini()` code to `illumos_crypto.c`

tcaputi commented:
@tonyhutter 

> I can't remember 100% why I moved some of the headers (it was months ago when I started on this...) but I think part of the reason was so they could be easily #included in the automated tests (like sha2_test.c).

That makes sense. For encryption, I used only used the generic interfaces so i didn't hit this problem. Another solution there could be to simply add the icp/include/ path to the test Makefile.

tuxoko commented:
@tonyhutter 
I don't think you should byteswap the checksum. If you look at the fletcher, the byteswap is for the data themself. The checksum value should always be native.

tonyhutter commented:
@tuxoko can you point to the code your referring to?  Are you talking about ld_swap* in edonr.c?

tonyhutter commented:
I've rebased the code and fixed the review comments:

- Added `*_fini()` to `edonr_mod/skein_mod.c`, removed `_info()`, added `*_init/*_fini()` to `illumos-crypto.c`
- Removed `skein_digest_update_mblk()/skein_digest_final_mblk()` from `skein_mod.c`
- Switch back to "`checksum = ZIO_CHECKSUM_OFF`" instead of `ZIO_CHECKSUM_NOPARITY` in `module/zfs/dmu.c`
- Removed "secret" from "Grab the secret checksum salt from the MOS" comment in spa.c
- Fixed "poool" spelling mistake in `libtest.shlib`
- Fixed edonr/skein compiler errors for `crypto_provider_info_t, crypto_ops_t, crypto_kcf_provider_handle_t`
- Used `&mod_cryptoops` instead of `&mod_miscops` in skein/edonr for `modlmisc` assignment
- Added `skein/edonr_mod.o` to `module/icp/Makefile.in`

Note that in order to get this code turned around quickly, I didn't do the the `cmn_err()` change (@behlendorf gave me the ok to skip it).  

Also, I talked to @tcaputi offline about https://github.com/zfsonlinux/zfs/pull/4760#r74871789 . He took another look at it, and doesn't think it's an issue after all.

tuxoko commented:
@tonyhutter 
I'm talking about `zio_checksum_SHA512_byteswap`, `zio_checksum_skein_byteswap` and `zio_checksum_edonr_byteswap`.

We should NOT do
```
	zcp->zc_word[0] = BSWAP_64(zcp->zc_word[0]);
	zcp->zc_word[1] = BSWAP_64(zcp->zc_word[1]);
	zcp->zc_word[2] = BSWAP_64(zcp->zc_word[2]);
	zcp->zc_word[3] = BSWAP_64(zcp->zc_word[3]);
```

And in fact, for byte based input hash like SHA (and I assume skein and edon-r are also byte based input), there should be a byteswap version. You should be able to see this in the lack of byteswap version in the SHA256.

tonyhutter commented:
Ok, I see what you're saying.  You're asking, why does sha512/skein/edonr use their "byteswap" function for their byteswap function pointer while sha256/fletcher use their "native" function for their byteswap function pointer:
```
/*
 * Information about each checksum function.
 */     
typedef struct zio_checksum_info {
        zio_checksum_t  *ci_func[2]; /* checksum function for each byteorder */
...


zio_checksum_info_t zio_checksum_table[ZIO_CHECKSUM_FUNCTIONS] = {
...
        {{zio_checksum_SHA256,          zio_checksum_SHA256},
            NULL, NULL, ZCHECKSUM_FLAG_METADATA | ZCHECKSUM_FLAG_DEDUP |
            ZCHECKSUM_FLAG_NOPWRITE, "sha256"},
...
        {{zio_checksum_SHA512_native,   zio_checksum_SHA512_byteswap},
            NULL, NULL, ZCHECKSUM_FLAG_METADATA | ZCHECKSUM_FLAG_DEDUP |
            ZCHECKSUM_FLAG_NOPWRITE, "sha512"},
...
        {{zio_checksum_skein_native,    zio_checksum_skein_byteswap},
            zio_checksum_skein_tmpl_init, zio_checksum_skein_tmpl_free,
            ZCHECKSUM_FLAG_METADATA | ZCHECKSUM_FLAG_DEDUP |
            ZCHECKSUM_FLAG_SALTED | ZCHECKSUM_FLAG_NOPWRITE, "skein"},
```
I don't know the answer.  I can only tell you that this is taken directly from illumos.  The native/byteswap functions seem to be called in two places:

1.`zio_checksum_compute()` always calls the "native" function directly (via ci->ci_func[0]).
2. `zio_checksum_error_impl()` may or may not call the "byteswap" function.

tuxoko commented:
@tonyhutter 
Like I said, if you look at `fletcher_4_byteswap`. It's doing bswap on **input data** not on zc_word. So what ever illumos is trying to do, it's not correct.

And the reason we don't byteswap zc_word should be obvious, because most of the time, zc_word is stored inside blkptr_t, meaning it will be swapped to native value when we read it into ARC. And in the case of ZIL blocks, as you can see in zio_checksum_error_impl, the "in-block checksum" `eck->zec_cksum` is swapped to native accordingly before comparison.

rlaager commented:
From what I can see, this has already been merged on Illumos, so if there is a byte-swapping mistake, fixing it would be an on-disk format breakage.

tuxoko commented:
@rlaager 
We cannot leave this issue because it will prevent importing pool from foreign endian machine. And I believe there shouldn't have any on-disk affect of this issue because we only write native endian value.

behlendorf commented:
@skiselkov can you take a minute look at @tuxoko's [byte swapping issue](https://github.com/zfsonlinux/zfs/pull/4760#issuecomment-250614451) above.  There's concern that this code @tonyhutter has ported directly from illumos us just wrong.

tuxoko commented:
@tonyhutter @behlendorf 
Ok, so I take a closer look at this, and I think I kind of know what the illumos is trying to do.

Originally in SHA256, we calculate the checksum and encode it into 64bit numbers. So no matter which endian you are, the same block will calculate into the same 64bit numbers(zc_word[]). And these numbers will be written to disk in the native endianness. When we import a foreign endian pool, the checksum will swapped to native endian as part of blkptr_t, or explicitly in `zio_checksum_error_impl`. So we don't need a separate function to calculate a block from foreign pool.

Now what they are trying to do with new checksum is, instead of encoding into 64bit numbers, it encode into byte stream. So no matter which endian you are, the same block will calculate to the same byte stream and written that byte stream to disk, but the zc_word[] will be reversed between different endian. Now when we import a foreign endian pool, the byte stream will still get byte swapped as 64bit numbers as part of blkptr_t and `zio_checksum_error_impl`. So to verify the blocks on foreign pool, we need a separate function version to swap the checksum byte stream calculated from foreign pool.

So this patch as it is would probably work fine when cross importing. However, this also raise the question how dedup would work when we do crosss import. As far as I can tell, dedup use checksum as in 64bit numbers(zc_word[]) as key to search. Which means you will not be able to find the entries in the dedup table written by the foreign machine even if you are writing the exact same data block.

And also, this does have on disk effect, so I'm not sure how we should deal with this.

behlendorf commented:
> So I'm not sure how we should deal with this.

We should merge this pull request as is to bring ZoL back in sync with OpenZFS.

Pools with these checksums already exist in the wild from illumos, so if there is an endianness issue lurking here which needs to be handled it can be tackled in another PR.  In practice big endian systems are fairly rare these days though so it's doubtful to cause users real problems.  Even modern ppc system and arm systems support little endian now.

As for the dedup issues you mentioned that's definitely something to address in a different PR.

behlendorf commented:
Thanks everybody I've merged this to master.
